### PR TITLE
Components: refactor `ToolsPanel` to pass `exhaustive-deps`

### PR DESCRIPTION
**Pull request by chad1008**
## What?
Updates the `ToolsPanel` component to satisfy the `exhaustive-deps` eslint rule

## Why?
Part of the effort in https://github.com/WordPress/gutenberg/pull/41166 to apply `exhuastive-deps` to the Components package

## How?
In `tools-panel/hook.ts`, we can safely remove `generateMenuItems` from dep arrays. React can't tell if an outer scope value like this changes, so it would never trigger a re-render, nor would it cause the effect to actually fire.

We can also safely remove `isResetting.current` from the dependency arrays. Including a `ref.current` value in a dependency array [isn't recommended in general](https://epicreact.dev/why-you-shouldnt-put-refs-in-a-dependency-array/).

In the relevant `useCallback`,  the value is updated, but never actually read... so even if it _had_ changed, that wouldn't impact the callback's behavior.

In the relevant `useMemo`, the value is read, so changes matter, but including the dependency in the array gives a false sense of security. If that were the only dep to change on a given render, React wouldn't notice, and the memoized value wouldn't update. If we really want the `isResetting` value to trigger a re-render to update this memo it should go into state, or a reducer. Because the component has functioned properly as it is now thus far, I think we should hold off on introducing new state (which will be sure to trigger additional renders and potentially other behavior changes) and simply remove the dependency. cc @aaronrobertshaw in case you have any insights, as you did some work with the `isResetting` value here most recently.

Finally, in `tools-panel-item/hook.ts`, a number of missing dependencies were added. In each case the added dep(s) are values being destructured from one of two contexts, so they'd cause their hooks to fire on every render. However, all of these hooks already have dependencies that are being destructured in this way, or derived from similar variables... so these new dependencies shouldn't change the existing behavior of anything. (Please feel free to correct me if my logic doesn't sound right here!)

## Testing Instructions
1. From your local Gutenberg directory, run `npx eslint --rule 'react-hooks/exhaustive-deps: warn' packages/components/src/tools-panel`
2. Confirm that the linter returns no errors
3. Confirm unit tests still pass
4. Run Storybook locally, confirm the components stories and/or docs still work as expected

### CODE REVIEW

**Review by aaronrobertshaw — APPROVED** [review]
Nice work @chad1008, appreciate the detailed notes and context 👍 

This tests well for me:

✅ No linting errors with `exhaustive-deps`
✅ Unit tests are passing
✅ Storybook examples continue to function
✅ No differences in behaviour in Block and Site Editors

> If we really want the isResetting value to trigger a re-render to update this memo it should go into state

I believe the `isResetting` value passed via context to the tools panel items was so that individual items could prevent calling their `onSelect`/`onDeselect` callbacks when the overall panel was executing a "reset all".  When the "reset all" is being done, the menu items state is updated which would force the `useMemo` update. So as you noted, we should be safe to simply remove the dependency.

File: packages/components/CHANGELOG.md
Comment (ciampo) [inline packages/components/CHANGELOG.md:22]: After merging #45043, I realized that its CHANGELOG entry didn't get added to the `Unreleased` section — would you be able to move that entry just under this one? Thank you!

Comment (chad1008) [reply packages/components/CHANGELOG.md:22]: oops! nice catch, I'm glad you spotted it. I've rebased this PR and moved the `Tooltip` CHANGELOG entry up to `Unreleased` 👍

**Comment by chad1008** [comment]
Thanks @aaronrobertshaw! Much appreciated!

@ciampo how are you feeling about this one? I particularly wanted to double check your thoughts on the added dependencies in `tools-panel-item/hook.ts`:

> ...all of these hooks already have dependencies that are being destructured in this way, or derived from similar variables... so these new dependencies shouldn't change the existing behavior of anything. (Please feel free to correct me if my logic doesn't sound right here!

If that sounds logical to you, I'll wait for the checks to pass and get this merged 🙂

**Comment by ciampo** [comment]
Changes look good to me — although I'm not sure I understand this part of your reasoning:

>  In each case the added dep(s) are values being destructured from one of two contexts, so they'd cause their hooks to fire on every render.

I may be wrong, but I don't think that destructuring values from those contexts automatically means that the hooks will run on every render.

---

Apart from that, I also gave a quick at how the component behaves in the editor and I couldn't spot any unexpected behavior. Plus, this component was mostly written by @aaronrobertshaw (including lots of unit tests!), and therefore seeing that he already approved this PR is great 🚀

**Pull request merged**
